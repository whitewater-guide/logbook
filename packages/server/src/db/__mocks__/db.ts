import {
  DatabasePoolConnectionType,
  DatabasePoolType,
  createPool,
} from 'slonik';

import { Client } from 'pg';

const cfg = {
  host: process.env.POSTGRES_HOST,
  user: process.env.POSTGRES_USER,
  database: `logbook_test_${process.env.JEST_WORKER_ID}`,
  password: process.env.POSTGRES_PASSWORD,
};

let _pool: DatabasePoolType | undefined;

export const db: () => DatabasePoolConnectionType = () => {
  if (!_pool) {
    throw new Error('no test pool found');
  }
  return _pool;
};

export const setupDB = async () => {
  const workerID = process.env.JEST_WORKER_ID;
  if (!workerID) {
    throw new Error('JEST_WORKER_ID not found');
  }
  const client = new Client({
    host: process.env.POSTGRES_HOST,
    user: process.env.POSTGRES_USER,
    database: process.env.POSTGRES_DB,
    password: process.env.POSTGRES_PASSWORD,
  });
  try {
    await client.connect();
    await client.query(`DROP DATABASE IF EXISTS logbook_test_${workerID}`);
    await client.query(
      `CREATE DATABASE logbook_test_${workerID} TEMPLATE logbook_test_template`,
    );
  } finally {
    await client.end();
  }

  _pool = createPool(
    `postgresql://${cfg.user}:${cfg.password}@${cfg.host}/${cfg.database}`,
  );
};

export const teardownDB = async () => {
  await _pool?.end();
};
