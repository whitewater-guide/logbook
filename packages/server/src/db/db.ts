import { createPool, DatabasePoolConnectionType } from 'slonik';

const cfg = {
  host: process.env.POSTGRES_HOST,
  user: process.env.POSTGRES_USER,
  database: process.env.POSTGRES_DB,
  password: process.env.POSTGRES_PASSWORD,
};

const pool = createPool(
  `postgresql://${cfg.user}:${cfg.password}@${cfg.host}/${cfg.database}`,
);

export const db: () => DatabasePoolConnectionType = () => pool;
export const setupDB = () => Promise.resolve();
export const teardownDB = async () => {
  await pool.end();
};
