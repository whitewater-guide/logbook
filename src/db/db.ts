import {
  DatabasePoolConnectionType,
  createPool,
  createTypeParserPreset,
} from 'slonik';

const cfg = {
  host: process.env.POSTGRES_HOST,
  user: process.env.POSTGRES_USER,
  database: process.env.POSTGRES_DB,
  password: process.env.POSTGRES_PASSWORD,
};

const typeParsers = createTypeParserPreset().filter(
  ({ name }) => !['name', 'timestamp', 'timestamptz'].includes(name),
);

const pool = createPool(
  `postgresql://${cfg.user}:${cfg.password}@${cfg.host}/${cfg.database}`,
  {
    typeParsers,
  },
);

export const db: () => DatabasePoolConnectionType = () => pool;
export const setupDB = () => Promise.resolve();
export const teardownDB = async () => {
  await pool.end();
};
