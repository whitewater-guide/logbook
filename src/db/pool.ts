import { createPool, createTypeParserPreset } from 'slonik';

const cfg = {
  host: process.env.POSTGRES_HOST,
  user: process.env.POSTGRES_USER,
  database:
    process.env.NODE_ENV === 'test'
      ? `logbook_test_${process.env.JEST_WORKER_ID}`
      : process.env.POSTGRES_DB,
  password: process.env.POSTGRES_PASSWORD,
};

const typeParsers = createTypeParserPreset().filter(
  ({ name }) => !['name', 'timestamp', 'timestamptz'].includes(name),
);

export const db = createPool(
  `postgresql://${cfg.user}:${cfg.password}@${cfg.host}/${cfg.database}`,
  {
    typeParsers,
  },
);
