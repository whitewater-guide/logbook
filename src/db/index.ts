import { db } from './pool';

export default db;
