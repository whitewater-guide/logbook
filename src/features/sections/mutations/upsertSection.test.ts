import {
  UpsertSectionMutation,
  UpsertSectionMutationVariables,
} from './upsertSection.test.generated';
import { setupDB, teardownDB } from '~/db';

import { SectionInput } from '~/__generated__/graphql';
import { USER_1 } from '~/test/fixtures';
import { gql } from 'apollo-server';
import { runQuery } from '~/test/apollo-helpers';

beforeEach(setupDB);
afterEach(teardownDB);

const mutation = gql`
  mutation upsertSection($section: SectionInput!) {
    upsertSection(section: $section) {
      id

      region
      river
      section
      difficulty
      putIn {
        lat
        lng
      }
      takeOut {
        lat
        lng
      }

      upstreamId
      upstreamData

      createdAt
      updatedAt
    }
  }
`;

const section: SectionInput = {
  region: 'Altai',
  river: 'Chuya',
  section: 'Mazhoy Gorge',
  difficulty: 5,
  putIn: { lat: 10, lng: 10 },
  takeOut: { lat: 20, lng: 20 },
  upstreamId: '__upstream_id__',
  upstreamData: { foo: 'bar ' },
};

it('should insert section', async () => {
  const result = await runQuery<
    UpsertSectionMutation,
    UpsertSectionMutationVariables
  >(mutation, { section }, USER_1);
  expect(result.errors).toBeUndefined();
  expect(result.data.upsertSection).toMatchSnapshot({
    createdAt: expect.any(Date),
    updatedAt: expect.any(Date),
    id: expect.any(String),
  });
});

it.todo('upsert with parent_id');
