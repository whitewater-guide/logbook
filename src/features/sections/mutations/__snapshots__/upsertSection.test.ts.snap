// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should insert section 1`] = `
Object {
  "createdAt": Any<Date>,
  "difficulty": 5,
  "id": Any<String>,
  "putIn": Object {
    "lat": 10,
    "lng": 10,
  },
  "region": "Altai",
  "river": "Chuya",
  "section": "Mazhoy Gorge",
  "takeOut": Object {
    "lat": 20,
    "lng": 20,
  },
  "updatedAt": Any<Date>,
  "upstreamData": Object {
    "foo": "bar ",
  },
  "upstreamId": "__upstream_id__",
}
`;
